src/vsr/journal.zig

```diff
@@ -1696,9 +1729,16 @@ pub fn Journal(comptime Replica: type, comptime Storage: type) type {
 
         fn write_prepare_release(self: *Self, write: *Self.Write, wrote: ?*Message) void {
             const replica = @fieldParentPtr(Replica, "journal", self);
-            write.callback(replica, wrote, write.trigger);
-            replica.message_bus.unref(write.message);
+            const write_callback = write.callback;
+            const write_trigger = write.trigger;
+            const write_message = write.message;
+
+            // Release the write prior to returning control to the caller.
+            // This allows us to enforce journal.writes.len≤1 when replica_count=1, because the
+            // callback may immediately start the next write.
             self.writes.release(write);
```

Comment:
Is there a scenario where releasing control to another caller ahead of the callback may invalidate other assertions in the callback? For example, what if the other caller changes the same op?

Reply:
> releasing control to another caller ahead of the callback

I'm not sure what you mean here — what other caller are you referring to?

Reply:
Another concurrent I/O task with it's own caller and callback. Does that make sense?

Reply:
I still don't understand.

```zig
        fn write_prepare_release(self: *Self, write: *Self.Write, wrote: ?*Message) void {
...

            self.writes.release(write); // ← (This used to be at the very end).
            write_callback(replica, wrote, write_trigger);
            replica.message_bus.unref(write_message);
}
```

- The `write_callback()` is being called before the end of `write_prepare_release`, so from the perspective of the journal context which calls `write_prepare_release`, nothing has changed.
- Within `write_callback` doesn't care about whether the write is released or not, except that an additional write might be available to use (this is the goal of the change).

`write_prepare_release` is called at the end of a callback chain to clean up, so I don't think there are any assertions we could collide with. But even if it wasn't, I don't think this change would affect the caller's invariants.

Reply:
Thanks @sentientwaffle — that's perfect.